sorting/patience_sort.c

```diff
@@ -0,0 +1,160 @@
+                flag--;
+                break;
+            }
+        }
+
+        if(flag) {
+            // The element in question can not be added to any existing piles, creating a new pile
+            piles[pileCount][0] = array[i];
+            pileCount++;
+        }
+    }
+
+    // This will keep track of the minimum value of all 'exposed' elements and which pile that value is from
+    int min, minLocation;
+
+    for (int i = 0; i < length; ++i) {
+        // Since there's no guarantee the first pile will be depleted slower than other piles,
+        // Example: when all elements are equal, in that case the first pile will be depleted immediately
+        // We can't simply initialize min to the top most element of the first pile,
+        // this loop finds a value to initialize min to.
+        for (int j = 0; j < pileCount; ++j) {
+            if(pileSizes[j] < 0) {
+                continue;
+            }
+            min = piles[j][pileSizes[j]];
+            minLocation = j;
+            break;
+        }
+
+        for (int j = 0; j < pileCount; ++j) {
+            if(pileSizes[j] < 0) {
+                continue;
+            }
+            if(piles[j][pileSizes[j]] < min) {
+                min = piles[j][pileSizes[j]];
+                minLocation = j;
+            }
+        }
+
+        array[i] = min;
+        pileSizes[minLocation]--;
+    }
+
+    // Deallocate memory
+    free(pileSizes);
+    for (int i = 0; i < length; ++i) {
+        free(piles[i]);
+    }
+    free(piles);
+}
+
+/**
+ * @brief Helper function to print an array
+ * @param array pointer to the array
+ * @param length length of the target array
+ * @returns void
+ */
+void printArray(int *array,int length) {
+    printf("Array:");
+    for (int i = 0; i < length; ++i) {
+        printf("%d",array[i]);
+        if (i != length - 1) putchar(',');
+    }
+    putchar('\n');
+}
+
+/**
+ * @brief Testing Helper function
+ * @param array pointer to the array to be used for testing
+ * @param length length of the target array
+ * @returns void
+ */
+
+void testArray(int *array,int length) {
+    printf("Before sorting:\n");
+    printArray(array,length);
+
+    patienceSort(array,length);
+
+    printf("After sorting:\n");
+    printArray(array,length);
+
+    for (int i = 0; i < length - 1; ++i) {
+        assert(array[i] <= array[i + 1]);
+    }
+    printf("All assertions have passed!\n\n");
+}
+
+/**
+ * @brief Self-test implementations
+ * @returns void
+ */
+static void test() {
+    int testArray1[] = {2,8,7,1,3,5,6,4};
+    int testArray2[] = {2,2,5,1,3,5,6,4};
+    int testArray3[] = {1,2,3,4,5,6,7,8};
+    int testArray4[] = {8,7,6,5,4,3,2,1};
+
+    testArray(testArray1,8);
+    testArray(testArray2,8);
+    testArray(testArray3,8);
+    testArray(testArray4,8);
+
+    printf("Testing successfully completed!\n");
+}
+
+/**
+ * @brief Main function
+ * @returns 0 on exit
+ */
+int main() {
+    test();
+    return 0;
+}
```

Comment:
Missing an endline. Without this endline, there could be issues with some IDEs and configurations.
```suggestion
}

```